packages/base-controller/CHANGELOG.md

```diff
@@ -7,27 +7,47 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-  - If the function parameters `allowedActions`, `allowedEvents` are a non-empty array, their corresponding type names must be explicitly passed into generic parameters `AllowedActions`, `AllowedEvents` to avoid type errors.
+- **BREAKING:** Narrow the return type of the function `deriveStateFromMetadata<ControllerState>` from `Record<string, Json>` to `Record<keyof ControllerState, Json>` ([#3949](https://github.com/MetaMask/core/pull/3949))
+  - TODO: What's the result of this change? `deriveStateFromMetadata` is not public, but `getAnonymizedState` and `getPersistentState` are.
+- **BREAKING:** Align type-level and runtime behavior of `getRestricted` so that omitted or empty inputs consistently represent a set of empty allowlists ([#4013](https://github.com/MetaMask/core/pull/4013))
+  - If the `AllowedActions` and `AllowedEvents` generic parameters are omitted, they are always assumed to be `never`.
+    - Previously, omission of these generic parameters resulted in the full allowlists for the controller being inferred as type constraints for the `allowedActions` and `allowedEvents` function parameters.
+  - If the function parameters `allowedActions` and `allowedEvents` are a non-empty array, their corresponding type names must be explicitly passed into generic parameters `AllowedActions` and `AllowedEvents` to avoid type errors.
     - This may cause some duplication of allowlists between type-level and value-level code.
     - This requirement is only relevant for TypeScript code. A JavaScript consumer only needs to pass in the correct value-level function parameters. Because of this, these changes should not affect downstream JavaScript code, but may be disruptive to TypeScript code.
-    - `getRestricted` is still able to flag `AllowedActions` and `AllowedEvents` members that should not be included in the allowlists, based on the `Action`, `Event` generic arguments passed into the `ControllerMessenger` instance.
-- **BREAKING:** The `RestrictedControllerMessenger` class constructor now expects `allowedActions` and `allowedEvents` as required options. ([$4013](https://github.com/MetaMask/core/pull/4013))
-- Convert interface `StatePropertyMetadata<T extends Json>` into a type.
+    - `getRestricted` is still able to flag `AllowedActions` and `AllowedEvents` members that should not be included in the allowlists, based on the `Action` and `Event` generic arguments passed into the `ControllerMessenger` instance.
+- **BREAKING:** The `RestrictedControllerMessenger` class constructor now expects `allowedActions` and `allowedEvents` as required options ([#4013](https://github.com/MetaMask/core/pull/4013))
+- **BREAKING**: Add `string` as generic constraint to the `Name` generic parameter of the types `NamespacedBy` and `NotNamespacedBy` ([#4036](https://github.com/MetaMask/core/pull/4036))
+  - TODO: `NamespacedBy` and `NotNamespacedBy` aren't public. What does this affect?
```

Comment:
@MajorLift What do you think?

Reply:
It looks like both `NamespacedBy` and `NotNamespacedBy` are package-level exports, since everything in `ControllerMessenger` is exported with a wildcard?

Reply:
Oh you're right! Removed in [5c424c0](https://github.com/MetaMask/core/pull/4039/commits/5c424c0d3fab25e706ffc1dc8e5e5e33fa03bd70).